Setup Wizard: Settings step

adekbadek opened the pull request:
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request, and the reason for such changes. -->

Settings step of the Setup Wizard.

Removes the `_newspack_profile_*` options, since they were not used anywhere anymore. Instead, WooCommerce `countrystate` & `currency` options are used. For this reason, WooCommerce is added as a required plugin during installation.  

Also, in a somewhat over-engineered way, this PR allows a component wrapped in a `withWizardScreen` HOC to render the wizard button. 

### How to test the changes in this Pull Request:

1. Install the plugin on a fresh WP site
2. Complete the first step (Settings) and verify the changes are saved: in Customizer, Reader Revenue wizard, and Yoast settings
3. Observe the site title in browser tab updating on "Site Title" input 😎

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

<!-- Mark completed items with an [x] -->

thomasguillot reviewed (changes requested):
I've pushed a bunch of changes.

One thing I'm not too sure about it the `ImageUpload` button. If you have a Site Icon, you click on "Replace" and then close the Modal, your Site Icon disappears.
Also, since we are requesting a square, it'd be nice to have a "Crop image" screen just like in the customizer.

For the social accounts, is it not possible to only have URLs and we somehow extract the twitter username for Yoast? Also, this doesn't seem to generate a Social Menu?

adekbadek commented:
> If you have a Site Icon, you click on "Replace" and then close the Modal, your Site Icon disappears.

Fixed in 5ddd8df8cb19fceca3386f5cb529b9211fdd38bf

> Also, since we are requesting a square, it'd be nice to have a "Crop image" screen just like in the customizer.

Added to https://github.com/Automattic/newspack-plugin/issues/852, the customizer flow is hard to reproduce here.

> For the social accounts, is it not possible to only have URLs and we somehow extract the twitter username for Yoast?

I'm not sure it's worth the added complexity - now we just mimic Yoast's settings.

> Also, this doesn't seem to generate a Social Menu?

No. Added to https://github.com/Automattic/newspack-plugin/issues/852

dkoo commented on assets/components/src/image-upload/index.test.js:
```diff
@@ -6,16 +6,6 @@ import ImageUpload from './';
-		const { getByAltText, getByText } = render( <ImageUpload image={ image } /> );
-		expect( getByAltText( 'Upload preview' ) ).toBeInTheDocument();
-		expect( getByText( 'Remove image' ) ).toBeInTheDocument();
+		expect( getByText( 'Upload' ) ).toBeInTheDocument();
```
Should we add a test for the remove button and background image when there's an image prop?

adekbadek replied:
👌  Added in 967f3b788d6558a421a5f182162719228d3b9dff

dkoo commented on assets/wizards/setup/views/settings/index.js:
```diff
@@ -8,48 +8,134 @@ import apiFetch from '@wordpress/api-fetch';
+		}
+	}, [ profileData.site_title ] );
+
+	const renderSetting = ( { options, label, key, type, placeholder, className } ) => {
```
Nit: why not make this a React component?

adekbadek replied:
Hm, it sort of is and isn't a real component - it's more of a switch statement lifted to a function. I'd keep it as it is.

dkoo commented on assets/wizards/setup/views/settings/index.js:
```diff
@@ -8,48 +8,134 @@ import apiFetch from '@wordpress/api-fetch';
 			/>
+			<Grid columns={ 3 } gutter={ 32 } rowGap={ 16 }>
+				{ wpseoFields.map( seoField => (
+					<Fragment key={ seoField.key }>
```
Is `Fragment` necessary since it only has one child?

adekbadek replied:
_Something_ has to have the `key` prop. The key could be passed to the component rendered by `renderSetting`, but I think it's cleaner this way.

dkoo replied:
Understood—I was thinking if `renderSetting` were a component, then that would get the `key`.

dkoo commented on includes/class-profile.php:
```diff
@@ -70,7 +151,7 @@ public static function newspack_get_option( $key ) {
 	 */
 	public static function newspack_update_option( $key, $value ) {
-		update_option( NEWSPACK_PROFILE_OPTION_PREFIX . $key, $value );
+		update_option( self::$opt_prefix . $key, $value );
```
Am I crazy or is `$opt_prefix` not defined anywhere?

adekbadek replied:
Forgot to remove this method in ef196e588835342758b0ac3b17057b0faa7e7d08, fixed in 967f3b788d6558a421a5f182162719228d3b9dff

dkoo reviewed (changes requested):
A couple of things:

Maybe an edge case, but if you reset a Newspack site that had been previously set up on an older version of the plugin, you might see the following for the social media links:

<img width="1067" alt="Screen Shot 2021-02-16 at 4 44 46 PM" src="https://user-images.githubusercontent.com/2230142/108135961-4c179d80-7076-11eb-986f-d83abccd3b33.png">

> WooCommerce `countrystate` & `currency` options are used. For this reason, WooCommerce is added as a required plugin during installation.

WooCommerce was removed as a default/required plugin in #726 because several Newspack sites who use NRH complained about the overhead of having to have WC installed. I get why you added it back, but it seems like it might irk those sites again. This could be overthinking it, but what if we did something like:

- If WooCommerce is installed already when going through Newspack setup, get the currency/location info from WooCommerce.
- If it isn't installed (and it probably won't be on new sites before they choose to install to support the Newspack donation platform), then we store these as an option as before.
- When activating WooCommerce, we check for the existence of the location/currency options and update WooCommerce settings to match.

adekbadek commented:
> Maybe an edge case, but if you reset a Newspack site that had been previously set up on an older version of the plugin, you might see the following for the social media links:

That's odd, since these fields were now part of the setup wizard before. These might be errors - can you please inspect the return payload of the `/newspack/v1/profile` call?

> WooCommerce was removed as a default/required plugin in #726 because several Newspack sites who use NRH complained about the overhead of having to have WC installed.

Sorry, my comment was imprecise. It's still not _a required plugin_ (as it was before #726), just installed on setup. If a user wishes not to use it, they can remove it afterwards and go with any other solution. The point here is providing defaults.

dkoo reviewed (approved):
> That's odd, since these fields were now part of the setup wizard before. These might be errors - can you please inspect the return payload of the /newspack/v1/profile call?

After advancing past the Settings screen, I can no longer replicate it, and I'm not seeing it in my local environment at all (which was obviously set up before checking out this branch), so let's defer until we see it happen again.

> Sorry, my comment was imprecise. It's still not a required plugin (as it was before #726), just installed on setup. If a user wishes not to use it, they can remove it afterwards and go with any other solution. The point here is providing defaults.

Ah, that makes sense to me. Thanks for clarifying.

adekbadek commented:
@thomasguillot Could you have another look?

thomasguillot commented:
@adekbadek all good for me 👍
